### ISSUE TITLE
Dask DataFrame API compatibility

### ISSUE THREAD

**Issue by TomAugspurger** [dask/dask#3688]
[This gist](https://gist.github.com/b9c20a55c996c68d0437e70459345f09) parses dask's and pandas dataframes. It produces two outputs

1. [Matching](https://docs.google.com/spreadsheets/d/1SD7AzJRxi6Dx8tblinIK52kNtiZnm0sv8yZOMxhLYMA/edit?usp=sharing)
    Parameters present in both implementations. These could be check for places where the default value or docstring doesn't match. Filter by `check_sig` and `check_doc`. 
2. [Differing](https://docs.google.com/spreadsheets/d/1qwUjS-LK4xV5EVjyh-vm15gKgw2xqfXUIGWmsEBAYms/edit?usp=sharing)
   Parameters where one is not present in the other. We would care about parameters that are not present in dask.

For example, here's a list of methods where the default values differ between dask and pandas:

method | parameter | pandas_value | dask_value
-- | -- | -- | --
min | skipna | skipna=True | skipna=None
idxmax | axis | axis=None | axis=0
where | other | other=nan | other=nan
nsmallest | n | n=5 | n
nsmallest | columns | columns=None | columns
drop | labels | labels | labels=None
std | skipna | skipna=True | skipna=None
max | skipna | skipna=True | skipna=None
sum | skipna | skipna=True | skipna=None
mask | other | other=nan | other=nan
prod | skipna | skipna=True | skipna=None
all | axis | axis=None | axis=0
idxmin | axis | axis=None | axis=0
nlargest | n | n=5 | n
nlargest | columns | columns=None | columns
eval | inplace | inplace=None | inplace=False
count | axis | axis=None | axis=0
any | axis | axis=None | axis=0
mean | skipna | skipna=True | skipna=None
corr | min_periods | min_periods=None | min_periods=1
var | skipna | skipna=True | skipna=None

Some questions:

1. Would this help guide a sprint?
2. If so, what other things should we inspect / highlight?

**Comment by mrocklin** [comment]
This looks great!  I think it would be very valuable for the sprint.  Would
a sensible goal be to add an automated test for this?

**Comment by TomAugspurger** [comment]
> Would a sensible goal be to add an automated test for this?

Oh, that's a good idea. I debated trashing all this, in despair that it would just fall out of date again, but a way to regularly check in would be helpful.

**Comment by TomAugspurger** [comment]
For the sprint, we could collect methods that are present on pandas.DataFrame but not dask.DataFrame. Maintainers could manually label the expected effort of implementing those methods.

**Comment by mrocklin** [comment]
How expensive would it be to repeat this process for dask array?

**Comment by jrbourbeau** [comment]
This looks super useful @TomAugspurger!

**Comment by TomAugspurger** [comment]
Should be easy to run for dask array. At this point, I'll clean up my hacky
script a bit.

**Comment by mrocklin** [comment]
Could also leave that task for the sprint itself

**Comment by TomAugspurger** [comment]
array-differ: https://docs.google.com/spreadsheets/d/1ydLNExcRm-1AM-Qb2hZZNyIu1NY-hJL2p1YumN5D_II/edit?usp=sharing
array-match: https://docs.google.com/spreadsheets/d/1dnmRISJ6SRZfWzIgZPx8Y006HSQbypWuXaS9Rtni46E/edit?usp=sharing

Refactoring the script and including it in the CI would be a useful sprint item.

**Comment by jakirkham** [comment]
Dask Array often excludes `copy` parameters, which may show up in this analysis.

**Comment by dharhas** [comment]
@TomAugspurger Is this still a valuable task. I've been looking for a good place to start contributing to Dask and this seems useful first step. I assume we need to do three things.

1. Clean up the code in your gist
2. Integrate into CI 
3. Possibly autogenerate a docs page?

Does that sound about right? Would we only test latest releases of Pandas and Dask?

**Comment by TomAugspurger** [comment]
It's probably still valuable, and cleaning up the script seems like a good place to start.

I'm not entirely sure how to integrate it into CI. Do you have thoughts? One option would be a simple text / JSON file somewhere in the repo with a list of known / OK differences, and if a difference is detected that isn't on that list then CI fails?

> Possibly autogenerate a docs page?

Do you mean a docs page that list "differences between pandas & dask"? That's interesting, but I'd be curious to see how large it is once we whittle down the unintentional differences. If it's not too large, then auto-generating that page would possibly be overkill.

And yes, I think only targeting the latest versions of pandas & NumPy makes sense.

**Comment by dharhas** [comment]
> I'm not entirely sure how to integrate it into CI. Do you have thoughts? One option would be a simple text / JSON file somewhere in the repo with a list of known / OK differences, and if a difference is detected that isn't on that list then CI fails?

Yes. Something like that sounds like the right approach. I would probably go with JSON/YAML in case we need a bit more metadata per difference.

> Do you mean a docs page that list "differences between pandas & dask"?

Yes, that was my original idea, but maybe it makes more sense to just finish out https://github.com/dask/dask/issues/1815

**Comment by Madhu94** [comment]
@TomAugspurger Is anyone still working on this? I'd like to help.  But doesn't this overlap with #1815 a bit? If that ticket was closed because we don't have this problem anymore, I'm wondering if this is still relevant.

**Comment by TomAugspurger** [comment]
No, but the data api group has some tooling that may replace the need for this: https://github.com/data-apis/array-api-comparison is the array one. There’s one for data frames too.
